README
	libanon - library of reusable anonymiztion functions.

  Copyright (C) 2008 Matus Harvan
  Copyright (C) 2008 Juergen Schoenwaelder

  libanon is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  This program is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with this program; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

Comments are welcome.

- Juergen Schoenwaelder <j.schoenwaelder@jacobs-university.de>


NEWS
Initial release of libanon, extracted from the snmpdump package in
order to turn this into a standalone software package.


autogen.sh
#!/bin/sh
#
# Run this to generate all the initial makefiles.
#
# $Id: autogen.sh 2030 2006-09-06 13:11:43Z schoenw $

DIE=true
PROJECT="libanon"

case `uname -s` in
    Darwin)
       libtool=glibtool
       libtoolize=glibtoolize
       ;;
    *)
       libtool=libtool
       libtoolize=libtoolize
       ;;
esac


VER=`autoconf --version | grep '^autoconf' | sed 's/.*)//'`
case "$VER" in
0* | 1\.* | 2\.[0-9] | 2\.[0-9][a-z]* | \
2\.1[0-2] | 2\.1[0-2][a-z]* )
  cat >&2 <<_EOF_

	You must have autoconf 2.13 or later installed to compile $PROJECT.
	Download the appropriate package for your distribution/OS,
	or get the source tarball at ftp://ftp.gnu.org/pub/gnu/autoconf/
_EOF_
  DIE="exit 1"
  ;;
esac


VER=`automake --version | grep '^automake' | sed 's/.*)//'`
case "$VER" in
0* | 1\.[0-3] | 1\.[0-3][a-z]* )

  cat >&2 <<_EOF_

	You must have automake 1.4 or later installed to compile $PROJECT.
	Download the appropriate package for your distribution/OS,
	or get the source tarball at ftp://ftp.gnu.org/pub/gnu/automake/
_EOF_
  DIE="exit 1"
  ;;
esac


VER=`${libtool} --version | grep ' libtool)' | \
sed 's/.*) \([0-9][0-9.]*\) .*/\1/' `
case "$VER" in
0* | 1\.[0-2] | 1\.[0-2][a-z]* | \
1\.3\.[0-2] | 1\.3\.[0-2][a-z]* )

  cat >&2 <<_EOF_

	You must have libtool 1.3.3 or later installed to compile $PROJECT.
	Download the appropriate package for your distribution/OS,
	or get the source tarball at ftp://ftp.gnu.org/pub/gnu/libtool/
_EOF_
  DIE="exit 1"
  ;;
esac

$DIE

${libtoolize} --copy --force || exit 1

if test -z "$*"; then
	echo "Running ./configure with no arguments. If you wish to pass any,"
	echo "please specify them on the $0 command line."
fi

#aclocal_flags="`./aclocal-flags`"
aclocal_flags=""
aclocalinclude="$ACLOCAL_FLAGS $aclocal_flags"; \
echo aclocal $aclocalinclude
aclocal $aclocalinclude || exit 1
echo autoheader
autoheader || exit 1
echo automake --add-missing --gnu $am_opt
automake --add-missing --gnu $am_opt || exit 1
echo autoconf
autoconf || exit 1

./configure "$@" || exit 1

echo
echo "Now type 'make' to compile $PROJECT."


src/xmalloc.c
/* xmalloc.c -- malloc with out of memory checking
   Copyright (C) 1990, 91, 92, 93, 94, 95, 96, 99 Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */

#if HAVE_CONFIG_H
# include <config.h>
#endif

#if __STDC__
# define VOID void
#else
# define VOID char
#endif

#include <stdio.h>		/* for stderr */

#if STDC_HEADERS

#include <sys/types.h>
#include <string.h>		/* for strlen etc. */
#include <stdlib.h>

#else  /* !STDC_HEADERS */

extern size_t strlen ();
extern char *strcpy ();

VOID *calloc ();
VOID *malloc ();
VOID *realloc ();
void free ();
#endif

#if ENABLE_NLS
# include <libintl.h>
# define _(Text) gettext (Text)
#else
# define textdomain(Domain)
# define _(Text) Text
#endif

#include <error.h>

/* Prototypes for functions defined here.  */
#if defined (__STDC__) && __STDC__
static VOID *fixup_null_alloc (size_t n);
VOID *xmalloc (size_t n);
VOID *xcalloc (size_t n, size_t s);
VOID *xrealloc (VOID *p, size_t n);
char *xstrdup (char *p);
#endif


static VOID *
fixup_null_alloc (n)
     size_t n;
{
  VOID *p;

  p = 0;
  if (n == 0)
    p = malloc ((size_t) 1);
  if (p == 0)
    {
      /* possible revisions: release some memory and re-try, print
	 more information (e.g. line number of input file) */
      fprintf(stderr, _("libanon: Memory exhausted"));
      exit(1);
    }
  return p;
}

/* Allocate N bytes of memory dynamically, with error checking.  */

VOID *
xmalloc (n)
     size_t n;
{
  VOID *p;

  p = malloc (n);
  if (p == 0)
    p = fixup_null_alloc (n);
  return p;
}

/* Allocate memory for N elements of S bytes, with error checking.  */

VOID *
xcalloc (n, s)
     size_t n, s;
{
  VOID *p;

  p = calloc (n, s);
  if (p == 0)
    p = fixup_null_alloc (n);
  return p;
}

/* Change the size of an allocated block of memory P to N bytes,
   with error checking.
   If P is NULL, run xmalloc.  */

VOID *
xrealloc (p, n)
     VOID *p;
     size_t n;
{
  if (p == 0)
    return xmalloc (n);
  p = realloc (p, n);
  if (p == 0)
    p = fixup_null_alloc (n);
  return p;
}

/* Make a copy of a string in a newly allocated block of memory. */

char *
xstrdup (str)
     char *str;
{
  VOID *p;

  p = xmalloc (strlen (str) + 1);
  strcpy (p, str);
  return p;
}


src/anon-mac.c
/*
 * anon-mac.c --
 *
 * IEEE MAC address anonymization functions.
 *
 * Broadcast addresses are preserved. Other multicast addresses are
 * anonymized, but the first bit is preserved.
 *
 * Copyright (c) 2005 Juergen Schoenwaelder
 */

#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <assert.h>
#include <openssl/lhash.h>
#include <openssl/rand.h>

#include "libanon.h"

#define MAC_LENGTH 6

/* node struct for a list */
struct node {
    uint8_t mac[MAC_LENGTH];
    struct node *next;
};

/* node struct for an lhash table */
struct hash_node {
    uint8_t mac[MAC_LENGTH];
    uint8_t hash[MAC_LENGTH];
};

/* For nonlexicographic order, we are generating hashes on the
 * fly. They are stored in anon_mac_t's list, to make sure we generate
 * unique addresses.
 *
 * For lexicographic order, we use anon_mac_t's list for storing the
 * real MAC addresses.
 */
struct _anon_mac {
    LHASH *hash_table;
    struct node *list;
    int state;
};

enum anon_mac_state_t {INIT=0, /* MAC anon object initialized,
				* set_used() may have been used
				* already, but no MAC address has yet
				* been anonymized using this object
				*/
		       NON_LEX, /* anon_mac_map() has already been used */
		       LEX}; /* anon_mac_map_lex() has already been used */



/* functions for the lhash table */
static unsigned long
anon_mac_hash(const struct hash_node *tohash)
{
    long hash = 0;
    int i;
    for(i=0;i<MAC_LENGTH;i++) {
	hash += tohash->mac[i];
	hash << 8;
    }
    return hash;
}

static int
anon_mac_cmp(const struct hash_node *arg1, const struct hash_node *arg2) {
    int i;
    for(i=0;i<MAC_LENGTH;i++) {
	if (arg1->mac[i] != arg2->mac[i]) return 1;
    }
    return 0;
}

/* Create the type-safe wrapper functions for use in the LHASH internals */
static IMPLEMENT_LHASH_HASH_FN(anon_mac_hash, const struct hash_node *);
static IMPLEMENT_LHASH_COMP_FN(anon_mac_cmp, const struct hash_node *);

/* returns 0 if we're trying to insert a MAC address not yet in the list */
static int
list_insert(struct node **list, const uint8_t *mac)
{
    struct node *p, *q, *n;
    int c;
    
    assert(mac);
    
    for (p = *list, q = NULL; p; q = p, p = p->next) {
	c = memcmp(mac, p->mac, MAC_LENGTH);
	if (c == 0) {
	    return 1;
	}
	if (c < 0) {
	    break;
	}
    }

    n = (struct node *) malloc(sizeof(struct node));
    if (! n) {
	return -1;
    }
    
    memset(n, 0, sizeof(struct node));
    memcpy(n->mac, mac, MAC_LENGTH);
    if (! q) {
	n->next = *list;
	*list = n;
    } else {
	n->next = q->next;
	q->next = n;
    }

    return 0;
}

/* 
 * check if MAC address is broadcast (ff:ff:ff:ff:ff:ff)
 */

static int
is_mac_broadcast(const uint8_t *mac)
{
  int i;
  for(i=0;i<MAC_LENGTH;i++) {
      if (mac[i] != 0xFF) {
	  return 0;
      }
  }
  return 1;
}

/*
 * Set/change the state of MAC anonymization object. Performs
 * neccessary checks if state change is ok.
 */
static int
anon_mac_set_state(anon_mac_t *a, int state)
{
    uint8_t amac[MAC_LENGTH];
    struct node *p, *q;

    assert(a);
    
    switch (state) {
    case INIT:
	if (a->state == state) return 0;
	break;
    case NON_LEX:
	if (a->state == state) return 0;
	assert(a->state == INIT);
	a->state = state;
	return 0;
    case LEX:
	if (a->state == state) return 0;
	assert(a->state == INIT);
	a->state = state;

	/* populate hashlist with unique random MAC addresses */
	struct node* hashlist = NULL;
	for (p = a->list; p; p = p->next) {
	    do {
		memset(amac,0,MAC_LENGTH);
		RAND_bytes(amac,MAC_LENGTH);
		/* RAND_pseudo_bytes(amac,6); */
		/* preserve first bit */
		if (p->mac[0] & 0xFF) {
		    amac[0] |= 0x80;
		    /* preserve broadcast */
		    if (is_mac_broadcast(p->mac)) {
			/* broadcast */
			amac[0] = amac[1] = amac[2] = amac[3]
			    = amac[4] = amac[5] = 0xFF;
		    } else {
			/* multicast */
			if (is_mac_broadcast(amac)) continue;
		    }
		} else {
		    /* unicast */
		    amac[0] &= 0x7F;
		}
	    } while (list_insert(&hashlist,amac)==1);
	}

	/* assign anon. macs to real macs in lhash table */
	struct hash_node *node;
	for (p = a->list, q = hashlist; p; q = q->next, p = p->next) {
	    node = (struct hash_node*) malloc(sizeof(struct hash_node));
	    assert(node);
	    memcpy(node->mac, p->mac, MAC_LENGTH);
	    memcpy(node->hash, q->mac, MAC_LENGTH);
	    lh_insert(a->hash_table, node);
	}

	/* we don't need the list of used MACs anymore */
	for (p = a->list; p; ) {
	    struct node *q = p;
	    p = p->next;
	    free(q);
	}
	a->list = NULL;
	/* we don't need the list of hashed MACs anymore */
	for (p = hashlist; p; ) {
	    struct node *q = p;
	    p = p->next;
	    free(q);
	}
	hashlist = NULL;

	return 0;
    default:
	fprintf(stderr,"trying to set ilegal state for an anon_mac_t\n");
	assert(0);
    }
    return 0;
}

/*
 * Create a new MAC anonymization object.
 */

anon_mac_t*
anon_mac_new()
{
    anon_mac_t *a;

    a = (anon_mac_t *) malloc(sizeof(anon_mac_t));
    if (! a) {
	return NULL;
    }
    memset(a, 0, sizeof(anon_mac_t));

    a->hash_table = lh_new(LHASH_HASH_FN(anon_mac_hash),
			   LHASH_COMP_FN(anon_mac_cmp));
    a->state = INIT; /* we're initializing, so we don't want to do
		      * checks on previous state values
		      */

    /* initialize randomness - this might not be the best way to do it */
    while (! RAND_status()) {
	fprintf(stderr, "initializing randomness...");
	char buf;
	buf = rand();
        RAND_seed(&buf, 1);
	fprintf(stderr, "done\n");
    }

    return a;
}

/*
 * recursively print list nodes
 */
static void
print(struct node *p)
{
    if (p) {
	fprintf(stderr, "[%p]\n", p);
	fprintf(stderr, "%02x:%02x:%02x:%02x:%02x:%02x\n",
		p->mac[0], p->mac[1], p->mac[2],
		p->mac[3], p->mac[4], p->mac[5]);
	print(p->next);
    }
}

/*
 * Delete an MAC anonymization object and free all its resources.
 */

void
anon_mac_delete(anon_mac_t *a)
{
    struct node *p;

    if (! a) {
	return;
    }

    lh_free(a->hash_table);

    for (p = a->list; p; ) {
	struct node *q = p;
	p = p->next;
	free(q);
    }

    free(a);
}

/*
 * Set the cryptographic key used for anonymization.
 */

void
anon_mac_set_key(anon_mac_t *a, const anon_key_t *key)
{
    assert(a);
    /* we might want to use the key to seed the RAND_* stuff */
}

/*
 * Mark a MAC address as used. We simply put it into a linked list
 * sorted by lexicographic order. Note that this is not really
 * efficient - a heap like data structure might be more appropriate.
 */

int
anon_mac_set_used(anon_mac_t *a, const uint8_t *mac)
{
    assert(a && mac);
    
    (void) anon_mac_set_state(a, INIT);

    if (list_insert(&(a->list), mac) >= 0) {
	return 0;
    } else{
	return -1;
    }
}

/*
 * anonymization on mac address
 * anonymized mac addresses are also kept in a linked list to make
 * sure they are unique
 */
int
anon_mac_map(anon_mac_t *a, const uint8_t *mac,
	     uint8_t *amac)
{
    struct hash_node node;
    struct hash_node *p;
    int tmp = 0;

    (void) anon_mac_set_state(a, NON_LEX);

    /* lookup anon. MAC in lhash table */
    memcpy(node.mac, mac, MAC_LENGTH);
    p = (struct hash_node *) lh_retrieve(a->hash_table,(void*) &node);
    
    if (p) { /* MAC found in lhash table */
	memcpy(amac, p->hash, MAC_LENGTH);
    } else { /* MAC not found in lhash table */
	/* generate a unique random MAC addresses */
	do {
	    memset(amac,0,MAC_LENGTH);
	    RAND_bytes(amac,MAC_LENGTH);
	    /* RAND_pseudo_bytes(amac,6); */
	    /* preserve first bit */
	    if (mac[0] & 0xFF) {
		amac[0] |= 0x80;
		/* preserve broadcast */
		if (is_mac_broadcast(mac)) {
		    /* broadcast */
		    amac[0] = amac[1] = amac[2] = amac[3]
			= amac[4] = amac[5] = 0xFF;
		} else {
		    /* multicast */
		    if (is_mac_broadcast(amac)) continue;
		}
	    } else {
		/* unicast */
		amac[0] &= 0x7F;
	    }
	    tmp = list_insert(&(a->list),amac);
	    assert(tmp >= 0);
	} while (tmp==1);
	/* store anon. MAC in lhash table */
	p = (struct hash_node*) malloc(sizeof(struct hash_node));
	assert(p);
	memcpy(p->mac, mac, MAC_LENGTH);
	memcpy(p->hash, amac, MAC_LENGTH);
	lh_insert(a->hash_table, p);
    }
    return 0;
}

/*
 * lexicographical-order-preserving anonymization on mac address
 */
int
anon_mac_map_lex(anon_mac_t *a, const uint8_t *mac,
		 uint8_t *amac)
{
    struct hash_node node;
    struct hash_node *p;
    (void) anon_mac_set_state(a, LEX);
    
    /* lookup the anonymized mac address in the lhash table */
    memset(&node, 0, sizeof(struct hash_node));
    memcpy(node.mac, mac, MAC_LENGTH);
    p = (struct hash_node *) lh_retrieve(a->hash_table,(void*) &node);
    assert(p);
    memcpy(amac, p->hash, MAC_LENGTH);
    return 0;
}


src/system.h
/* system-dependent definitions for libanon

   Written for GNU fileutils.
   Adapted for autoproject by James R. Van Zandt.

   Copyright (C) 89, 91, 92, 93, 94, 1996 Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software Foundation,
   Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.  */

#if HAVE_CONFIG_H
#include "config.h"
#endif /* HAVE_CONFIG_H */

/* Include sys/types.h before this file.  */

#include <sys/stat.h>

#ifdef STAT_MACROS_BROKEN
#undef S_ISBLK
#undef S_ISCHR
#undef S_ISDIR
#undef S_ISFIFO
#undef S_ISLNK
#undef S_ISMPB
#undef S_ISMPC
#undef S_ISNWK
#undef S_ISREG
#undef S_ISSOCK
#endif /* STAT_MACROS_BROKEN.  */

#ifndef S_IFMT
#define S_IFMT 0170000
#endif
#if !defined(S_ISBLK) && defined(S_IFBLK)
#define S_ISBLK(m) (((m) & S_IFMT) == S_IFBLK)
#endif
#if !defined(S_ISCHR) && defined(S_IFCHR)
#define S_ISCHR(m) (((m) & S_IFMT) == S_IFCHR)
#endif
#if !defined(S_ISDIR) && defined(S_IFDIR)
#define S_ISDIR(m) (((m) & S_IFMT) == S_IFDIR)
#endif
#if !defined(S_ISREG) && defined(S_IFREG)
#define S_ISREG(m) (((m) & S_IFMT) == S_IFREG)
#endif
#if !defined(S_ISFIFO) && defined(S_IFIFO)
#define S_ISFIFO(m) (((m) & S_IFMT) == S_IFIFO)
#endif
#if !defined(S_ISLNK) && defined(S_IFLNK)
#define S_ISLNK(m) (((m) & S_IFMT) == S_IFLNK)
#endif
#if !defined(S_ISSOCK) && defined(S_IFSOCK)
#define S_ISSOCK(m) (((m) & S_IFMT) == S_IFSOCK)
#endif
#if !defined(S_ISMPB) && defined(S_IFMPB) /* V7 */
#define S_ISMPB(m) (((m) & S_IFMT) == S_IFMPB)
#define S_ISMPC(m) (((m) & S_IFMT) == S_IFMPC)
#endif
#if !defined(S_ISNWK) && defined(S_IFNWK) /* HP/UX */
#define S_ISNWK(m) (((m) & S_IFMT) == S_IFNWK)
#endif

#ifndef S_IEXEC
#define S_IEXEC S_IXUSR
#endif

#ifndef S_IXUSR
#define S_IXUSR S_IEXEC
#endif
#ifndef S_IXGRP
#define S_IXGRP (S_IEXEC >> 3)
#endif
#ifndef S_IXOTH
#define S_IXOTH (S_IEXEC >> 6)
#endif
#ifndef S_IXUGO
#define S_IXUGO (S_IXUSR | S_IXGRP | S_IXOTH)
#endif

#if !defined(HAVE_MKFIFO)
#define mkfifo(path, mode) (mknod ((path), (mode) | S_IFIFO, 0))
#endif

#ifdef HAVE_SYS_PARAM_H
#include <sys/param.h>
#endif

/* <unistd.h> should be included before any preprocessor test
   of _POSIX_VERSION.  */
#ifdef HAVE_UNISTD_H
#include <unistd.h>
#endif

#ifndef STDIN_FILENO
#define STDIN_FILENO 0
#endif

#ifndef STDOUT_FILENO
#define STDOUT_FILENO 1
#endif

#ifndef STDERR_FILENO
#define STDERR_FILENO 2
#endif

/* FIXME: Don't use _POSIX_VERSION.  */
#ifndef _POSIX_VERSION
off_t lseek ();
#endif

#ifdef TM_IN_SYS_TIME
#include <sys/time.h>
#else
#include <time.h>
#endif

/* Since major is a function on SVR4, we can't use `ifndef major'.  */
#ifdef MAJOR_IN_MKDEV
#include <sys/mkdev.h>
#define HAVE_MAJOR
#endif
#ifdef MAJOR_IN_SYSMACROS
#include <sys/sysmacros.h>
#define HAVE_MAJOR
#endif
#ifdef major			/* Might be defined in sys/types.h.  */
#define HAVE_MAJOR
#endif

#ifndef HAVE_MAJOR
#define major(dev)  (((dev) >> 8) & 0xff)
#define minor(dev)  ((dev) & 0xff)
#define makedev(maj, min)  (((maj) << 8) | (min))
#endif
#undef HAVE_MAJOR

#ifdef HAVE_UTIME_H
#include <utime.h>
#endif

/* Some systems (even some that do have <utime.h>) don't declare this
   structure anywhere.  */
#ifndef HAVE_STRUCT_UTIMBUF
struct utimbuf
{
  long actime;
  long modtime;
};
#endif

/* Don't use bcopy!  Use memmove if source and destination may overlap,
   memcpy otherwise.  */

#ifdef HAVE_STRING_H
# if !STDC_HEADERS && HAVE_MEMORY_H
#  include <memory.h>
# endif
# include <string.h>
#else
# include <strings.h>
char *memchr ();
#endif

#include <errno.h>
#ifndef errno
extern int errno;
#endif

#ifdef STDC_HEADERS
#define getopt system_getopt
#include <stdlib.h>
#undef getopt
#else /* not STDC_HEADERS */
char *getenv ();
#endif /* STDC_HEADERS */

/* The following test is to work around the gross typo in
   systems like Sony NEWS-OS Release 4.0C, whereby EXIT_FAILURE
   is defined to 0, not 1.  */
#if !EXIT_FAILURE
# undef EXIT_FAILURE
# define EXIT_FAILURE 1
#endif

#ifndef EXIT_SUCCESS
# define EXIT_SUCCESS 0
#endif

#ifdef HAVE_FCNTL_H
#include <fcntl.h>
#else
#include <sys/file.h>
#endif

#ifndef SEEK_SET
#define SEEK_SET 0
#define SEEK_CUR 1
#define SEEK_END 2
#endif
#ifndef F_OK
#define F_OK 0
#define X_OK 1
#define W_OK 2
#define R_OK 4
#endif

#ifdef HAVE_DIRENT_H
# include <dirent.h>
# define NLENGTH(direct) (strlen((direct)->d_name))
#else /* not HAVE_DIRENT_H */
# define dirent direct
# define NLENGTH(direct) ((direct)->d_namlen)
# ifdef HAVE_SYS_NDIR_H
#  include <sys/ndir.h>
# endif /* HAVE_SYS_NDIR_H */
# ifdef HAVE_SYS_DIR_H
#  include <sys/dir.h>
# endif /* HAVE_SYS_DIR_H */
# ifdef HAVE_NDIR_H
#  include <ndir.h>
# endif /* HAVE_NDIR_H */
#endif /* HAVE_DIRENT_H */

#ifdef CLOSEDIR_VOID
/* Fake a return value. */
#define CLOSEDIR(d) (closedir (d), 0)
#else
#define CLOSEDIR(d) closedir (d)
#endif

/* Get or fake the disk device blocksize.
   Usually defined by sys/param.h (if at all).  */
#ifndef DEV_BSIZE
#ifdef BSIZE
#define DEV_BSIZE BSIZE
#else /* !BSIZE */
#define DEV_BSIZE 4096
#endif /* !BSIZE */
#endif /* !DEV_BSIZE */

/* Extract or fake data from a `struct stat'.
   ST_BLKSIZE: Optimal I/O blocksize for the file, in bytes.
   ST_NBLOCKS: Number of 512-byte blocks in the file
   (including indirect blocks). */
#ifndef HAVE_ST_BLOCKS
# define ST_BLKSIZE(statbuf) DEV_BSIZE
# if defined(_POSIX_SOURCE) || !defined(BSIZE) /* fileblocks.c uses BSIZE.  */
#  define ST_NBLOCKS(statbuf) (((statbuf).st_size + 512 - 1) / 512)
# else /* !_POSIX_SOURCE && BSIZE */
#  define ST_NBLOCKS(statbuf) (st_blocks ((statbuf).st_size))
# endif /* !_POSIX_SOURCE && BSIZE */
#else /* HAVE_ST_BLOCKS */
/* Some systems, like Sequents, return st_blksize of 0 on pipes. */
# define ST_BLKSIZE(statbuf) ((statbuf).st_blksize > 0 \
			       ? (statbuf).st_blksize : DEV_BSIZE)
# if defined(hpux) || defined(__hpux__) || defined(__hpux)
/* HP-UX counts st_blocks in 1024-byte units.
   This loses when mixing HP-UX and BSD filesystems with NFS.  */
#  define ST_NBLOCKS(statbuf) ((statbuf).st_blocks * 2)
# else /* !hpux */
#  if defined(_AIX) && defined(_I386)
/* AIX PS/2 counts st_blocks in 4K units.  */
#   define ST_NBLOCKS(statbuf) ((statbuf).st_blocks * 8)
#  else /* not AIX PS/2 */
#   if defined(_CRAY)
#    define ST_NBLOCKS(statbuf) ((statbuf).st_blocks * ST_BLKSIZE(statbuf)/512)
#   else /* not AIX PS/2 nor CRAY */
#    define ST_NBLOCKS(statbuf) ((statbuf).st_blocks)
#   endif /* not _CRAY */
#  endif /* not AIX PS/2 */
# endif /* !hpux */
#endif /* HAVE_ST_BLOCKS */

/* Convert B 512-byte blocks to kilobytes if K is nonzero,
   otherwise return it unchanged. */
#define convert_blocks(b, k) ((k) ? ((b) + 1) / 2 : (b))

#ifndef RETSIGTYPE
#define RETSIGTYPE void
#endif

#ifdef __GNUC__
# undef alloca
# define alloca __builtin_alloca
#else
# ifdef HAVE_ALLOCA_H
#  include <alloca.h>
# else
#  ifndef _AIX
/* AIX alloca decl has to be the first thing in the file, bletch! */
char *alloca ();
#  endif
# endif
#endif

#include <ctype.h>

/* Jim Meyering writes:

   "... Some ctype macros are valid only for character codes that
   isascii says are ASCII (SGI's IRIX-4.0.5 is one such system --when
   using /bin/cc or gcc but without giving an ansi option).  So, all
   ctype uses should be through macros like ISPRINT...  If
   STDC_HEADERS is defined, then autoconf has verified that the ctype
   macros don't need to be guarded with references to isascii. ...
   Defining isascii to 1 should let any compiler worth its salt
   eliminate the && through constant folding."

   Bruno Haible adds:

   "... Furthermore, isupper(c) etc. have an undefined result if c is
   outside the range -1 <= c <= 255. One is tempted to write isupper(c)
   with c being of type `char', but this is wrong if c is an 8-bit
   character >= 128 which gets sign-extended to a negative value.
   The macro ISUPPER protects against this as well."  */

#if defined (STDC_HEADERS) || (!defined (isascii) && !defined (HAVE_ISASCII))
#define IN_CTYPE_DOMAIN(c) 1
#else
#define IN_CTYPE_DOMAIN(c) isascii(c)
#endif

#ifdef isblank
#define ISBLANK(c) (IN_CTYPE_DOMAIN (c) && isblank (c))
#else
#define ISBLANK(c) ((c) == ' ' || (c) == '\t')
#endif
#ifdef isgraph
#define ISGRAPH(c) (IN_CTYPE_DOMAIN (c) && isgraph (c))
#else
#define ISGRAPH(c) (IN_CTYPE_DOMAIN (c) && isprint (c) && !isspace (c))
#endif

#define ISPRINT(c) (IN_CTYPE_DOMAIN (c) && isprint (c))
#define ISALNUM(c) (IN_CTYPE_DOMAIN (c) && isalnum (c))
#define ISALPHA(c) (IN_CTYPE_DOMAIN (c) && isalpha (c))
#define ISCNTRL(c) (IN_CTYPE_DOMAIN (c) && iscntrl (c))
#define ISLOWER(c) (IN_CTYPE_DOMAIN (c) && islower (c))
#define ISPUNCT(c) (IN_CTYPE_DOMAIN (c) && ispunct (c))
#define ISSPACE(c) (IN_CTYPE_DOMAIN (c) && isspace (c))
#define ISUPPER(c) (IN_CTYPE_DOMAIN (c) && isupper (c))
#define ISXDIGIT(c) (IN_CTYPE_DOMAIN (c) && isxdigit (c))
#define ISDIGIT_LOCALE(c) (IN_CTYPE_DOMAIN (c) && isdigit (c))

/* ISDIGIT differs from ISDIGIT_LOCALE, as follows:
   - Its arg may be any int or unsigned int; it need not be an unsigned char.
   - It's guaranteed to evaluate its argument exactly once.
   - It's typically faster.
   Posix 1003.2-1992 section 2.5.2.1 page 50 lines 1556-1558 says that
   only '0' through '9' are digits.  Prefer ISDIGIT to ISDIGIT_LOCALE unless
   it's important to use the locale's definition of `digit' even when the
   host does not conform to Posix.  */
#define ISDIGIT(c) ((unsigned) (c) - '0' <= 9)

#ifndef __P
#if defined (__GNUC__) || (defined (__STDC__) && __STDC__)
#define __P(args) args
#else
#define __P(args) ()
#endif  /* GCC.  */
#endif  /* Not __P.  */

/* Take care of NLS matters.  */

#if HAVE_LOCALE_H
# include <locale.h>
#endif
#if !HAVE_SETLOCALE
# define setlocale(Category, Locale) /* empty */
#endif

#if ENABLE_NLS
# include <libintl.h>
# define _(Text) gettext (Text)
#else
# undef bindtextdomain
# define bindtextdomain(Domain, Directory) /* empty */
# undef textdomain
# define textdomain(Domain) /* empty */
# define _(Text) Text
#endif

#define STREQ(a,b) (strcmp((a), (b)) == 0)
